### DISCUSSION TITLE
Deserialize array of objects in a property of an object

### DISCUSSION THREAD

**Discussion by Crell** [symfony/symfony#42467] [Q&A] [answered]
Serializer component question.  I have an array like this:

```php
$data =  [
        'caching' => [
            'cacheConfigurations' => [
                'hash' => [
                    'backend' => 'TYPO3\\CMS\\Core\\Cache\\Backend\\Typo3DatabaseBackend',
                ],
                'imagesizes' => [
                    'backend' => 'TYPO3\\CMS\\Core\\Cache\\Backend\\Typo3DatabaseBackend',
                    'options' => [
                        'compression' => true,
                    ],
                ],
           ],
       ],
];
```

And I'm trying to denormalize it into an object structure like this:

```php
class System
{
    public function __construct(
        /** @var CacheConfig[] */
        public array $cacheConfigurations = [],
}

class CacheConfig 
{
    public function __construct(
        public string $backend = '',
        #[SerializedName('options')]
        public ?CacheOptions $options = null,
    ) {}
}
```

Serializer definition:

```php
    protected function getSerializer(): Serializer
    {
        $classMetadataFactory = new ClassMetadataFactory(new AnnotationLoader(new AnnotationReader()));
        $metadataAwareNameConverter = new MetadataAwareNameConverter($classMetadataFactory);
        $infoExtractor = new ReflectionExtractor();

        $normalizer =  new ObjectNormalizer(
            classMetadataFactory: $classMetadataFactory,
            nameConverter: $metadataAwareNameConverter,
            propertyTypeExtractor: $infoExtractor,
        );

        return new Serializer([new DateTimeNormalizer(), new PropertyNormalizer(), new ArrayDenormalizer(), $normalizer]);
    }
```

Problem: The `$cacheConfigurations` property is always coming back as an empty array.  It's just getting skipped.  I do have the `ArrayDenormalizer` set up in my Serializer class, but I'm not clear on how to trigger it in this case.  Suggestions?  What connection or definition am I missing?

**Comment by lyrixx** [comment, marked as answer]
I guess you missed the PHPdoc extractor

**Reply by Crell** [reply]
Ah, that's a thing, is it.  Interesting.  Does that also imply reflection, though?  (99% of the time reflection is best, followed by attributes, followed by docblocks as the least favorable option.)

**Reply by lyrixx** [reply]
Yes, but you are using PHPdoc 👍

**Reply by Crell** [reply]
In this particular case, yes.  But if there is a normal type declared, I want to use that.  That means using the ReflectionExtractor, because I definitely cannot tell my users to double-enter property definitions most of the time.  (Oh to have Generics...)  I tested switching to the PhpDocExtractor, and it then failed on both this code *and* another class I am deserializing that does have a normal typed object property.  So, that seems like it should work for this example but didn't.

How can I have multiple extractors running?

**Reply by lyrixx** [reply]
I think you can try something like: https://github.com/symfony/symfony/blob/1d196159fde010590b857059b5219922031c284b/src/Symfony/Component/Serializer/Tests/Normalizer/ObjectNormalizerTest.php#L539-L546

**Reply by Crell** [reply]
Hm.  That allows the typed properties to work again, but the array-of-objects property is still not being populated.

**Reply by lyrixx** [reply]
Could you paste a full reproducer please,  🙏🏼 ?

**Reply by Crell** [reply]
Code here: https://github.com/Crell/serializer-test

It's a phpunit project, just testing out the serializer, basically.  The specific test that's failing is this:

https://github.com/Crell/serializer-test/blob/master/tests/SerializerConfigTest.php#L127

Where I'm trying to objectify this segment of a config file: https://github.com/Crell/serializer-test/blob/master/tests/LocalConfiguration.php#L85

Everything except the cacheConfigurations part is working.  That one, I'm just not sure how to coax it into loading the items into CacheConfig objects.  Instead the var_dump() shows it's always just an empty array.

(Note: There is a non-zero chance I'm just making some incredibly stupid embarrassing mistake.  I am prepared for this.)

**Reply by lyrixx** [reply]
You have a mismatch between the PHP model and the raw data. See the PR, it'll be simpler to understand :) 

https://github.com/Crell/serializer-test/pull/1

---
You code was supposed to process:
```php
    'SYS' => [
        'cacheConfigurations' => [
```
But it's:
```php
    'SYS' => [
        'caching' => [
            'cacheConfigurations' => [
```

**Reply by Crell** [reply]
Ahhh...  Yep, incredibly stupid embarrassing mistake it was.  Thanks!
